packages/shell-api/src/shell-internal-state.ts

```diff
@@ -0,0 +1,238 @@
+    this.connectionInfo = await this.currentDb.mongo.serviceProvider.getConnectionInfo();
+    this.messageBus.emit('mongosh:connect', this.connectionInfo.extraInfo);
+  }
+
+  // TODO: do we need to close all Mongos ever made?
+  close(p): void {
+    try {
+      this.currentDb.mongo.serviceProvider.close(p);
+      // eslint-disable-next-line no-empty
+    } catch (e) {
+    }
+  }
+
+  use(db): any {
+    return this.currentDb.mongo.use(db);
+  }
+  show(arg): any {
+    return this.currentDb.mongo.show(arg);
+  }
+  async it(): Promise<any> {
+    if (!this.currentCursor) {
+      // TODO: warn here
+      return new CursorIterationResult();
+    }
+    return await this.currentCursor._it();
+  }
+  public help(): Help {
+    this.messageBus.emit('mongosh:help');
+    return new Help({
+      help: 'shell-api.help.description',
+      docs: 'https://docs.mongodb.com/manual/reference/method',
+      attr: [
+        {
+          name: 'use',
+          description: 'shell-api.help.help.use'
+        },
+        {
+          name: 'it',
+          description: 'shell-api.help.help.it'
+        },
+        {
+          name: 'show databases',
+          description: 'shell-api.help.help.show-databases'
+        },
+        {
+          name: 'show collections',
+          description: 'shell-api.help.help.show-collections'
+        },
+        {
+          name: '.exit',
+          description: 'shell-api.help.help.exit'
+        }
+      ]
+    });
+  }
+
+  /**
+   * Prepare a `contextObject` as global context and set it as context
+   * Add each attribute to the AsyncWriter also.
+   *
+   * The `contextObject` is prepared so that it can be used as global object
+   * for the repl evaluationi.
+   *
+   * @note The `contextObject` is mutated, it will retain all of its existing
+   * properties but also have the global shell api objects and functions.
+   *
+   * @param {Object} contextObject - contextObject an object used as global context.
+   */
+  setCtx(contextObject: any): void {
+    // Add API methods for VSCode and scripts
+    contextObject.use = this.use.bind(this);
+    contextObject.show = this.show.bind(this);
+    contextObject.it = this.it.bind(this);
+    contextObject.help = this.help.bind(this);
+    contextObject.toIterator = toIterator;
+    contextObject.print = async(arg) => {
+      if (arg.toReplString) {
+        console.log(await arg.toReplString());
+      } else {
+        console.log(arg);
+      }
+    };
+    contextObject.printjson = contextObject.print;
+    Object.assign(contextObject, ShellBson);
+    contextObject.rs = new ReplicaSet(this.currentDb.mongo);
+    contextObject.sh = new Shard(this.currentDb.mongo);
+
+    // Add global shell objects
+    this.asyncWriter.symbols.initializeApiObjects({
+      db: signatures.Database,
+      rs: signatures.ReplicaSet,
+      sh: signatures.Shard,
+      Mongo: signatures.Mongo
+    });
+
+    try {
+      Object.defineProperty(contextObject, 'db', {
```

Comment:
this would look so much easier if we extract it in a class

Reply:
I don't know what you mean by that, can you write an example?

Reply:
Ideally I'd love to have all the global variables and methods in the same place, as class accessors and methods, but I understand is probably impossible or it wont fit in the rest of the design. Something like this:

``` js
class ShellApiGlobals {
  private _currentDb: Database;
  private _replicaSetCache: Map<Mongo, ReplicaSet>;

  constructor() {
    this._currentDb = new Database(/** ... */);
    this._replicaSetCache = new Map<Mongo, ReplicaSet>();
  }
  
  get db(): Database {
    return this._currentDb;
  }

  set db(database: Database) {
    this._currentDb = database;
  }

  get rs() {
    if (!this._replicaSetCache.has(this._currentDb.mongo)) {
      this._replicaSetCache.set(new ReplicaSet(this._currentDb.mongo));
    }

    return this._replicaSetCache.get(this._currentDb);
  }
}
```

So that `setCtx` would be just like this:

``` js
class ShellInternalState {
  constructor() {
    this.shellApiGlobals = new ShellApiGlobals(/** ... */);
  }

  setCtx(contextObject) {
    const globalAttributes = Object.keys(signatures.ShellApiGlobals.attributes);

    for (const attributeName of globalAttributes) {
      contextObject[attributeName] = typeof this.shellApiGlobals[attributeName] === 'function' ?
        this.shellApiGlobals[attributeName].bind(this.shellApiGlobals) :
        this.shellApiGlobals[attributeName]
    }
  }
}
```

Totally fine to keep this for later.

Reply:
https://jira.mongodb.org/browse/MONGOSH-216